### PR TITLE
Implement deadlock detection

### PR DESCRIPTION
**Pull request by arthurprs**
My stab to solve cc #18 using a wait graph. It requires more work but overall I think it's a good candidate for inclusion. Let me know your thoughts.

[x] feature flagged
[x] reasonably efficient
[nop] checking thread (automatic start?)
[x] reporting (how to?)
[ ] locking/unlocking in different threads break the algorithm

### CODE REVIEW

**Comment by arthurprs** [comment]
I need to figure out how to test this properly.

**Comment by arthurprs** [comment]
Ok, it's in good shape for review now.

File: core/src/parking_lot.rs [hunk 1]
```diff
@@ -1035,3 +1042,67 @@ unsafe fn unpark_filter_internal(key: usize,
...
+            },
+        };
+    }
+}
+
+/// Returns true if a deadlock is detected
+#[cfg(feature = "deadlock_detection")]
+pub fn check_deadlock() -> bool {
+    // basic algorithm that builds a directed graph of the threads and the resources
+    // then checks for the presence of cycles (deadlocks)
```
Comment (Amanieu) [inline core/src/parking_lot.rs]: While checking for deadlocks, you should lock the entire hash table, not just one bucket at a time. See `grow_hashtable` for how to do this properly. You might get false positives if you do not do this.

File: core/src/parking_lot.rs [hunk 2]
```diff
@@ -1035,3 +1042,67 @@ unsafe fn unpark_filter_internal(key: usize,
...
+#[cfg(feature = "deadlock_detection")]
+pub fn deregister_resource(key: usize) {
+    unsafe {
+        let mut thread_data = None;
+        let thread_data = get_thread_data(&mut thread_data);
+        let resources = &mut (*thread_data.resources.get());
+        match resources.iter().rposition(|x| *x == key) {
+            Some(p) => resources.swap_remove(p),
+            None => {
+                check_deadlock();
```
Comment (Amanieu) [inline core/src/parking_lot.rs]: Why do you call `check_deadlock` here? This code can only be reached if there is a bug in the implementation of `Mutex` (or whatever primitive calls `deregister_resource`).

File: core/src/parking_lot.rs [hunk 3]
```diff
@@ -132,6 +135,9 @@ struct ThreadData {
 
     // ParkToken value set by the thread when it was parked
     park_token: Cell<ParkToken>,
+
+    // Resources
+    resources: UnsafeCell<Vec<usize>>,
```
Comment (Amanieu) [inline core/src/parking_lot.rs]: The name "resources" isn't very informative. You should make it clearer that this is for deadlock detection.

File: core/src/lib.rs [hunk 4]
```diff
@@ -77,3 +80,5 @@ pub use parking_lot::{ParkResult, UnparkResult, RequeueOp, UnparkToken, ParkToke
 pub use parking_lot::{DEFAULT_UNPARK_TOKEN, DEFAULT_PARK_TOKEN};
 pub use parking_lot::{park, unpark_one, unpark_all, unpark_requeue, unpark_filter};
 pub use spinwait::SpinWait;
+#[cfg(feature = "deadlock_detection")]
+pub use parking_lot::{register_resource, deregister_resource, check_deadlock};
```
Comment (Amanieu) [inline core/src/lib.rs]: I think the names `register_resource` and `deregister_resource` aren't very informative: it isn't immediately clear that this is related to deadlock detection. Maybe you could put these in their own submodule `parking_lot_core::deadlock_detector`.

Comment (Amanieu) [inline core/src/lib.rs]: It would also be nice if these functions we no-ops if the `deadlock_detection` feature is not defined. This avoids the need to spam `#[cfg(feature = "deadlock_detection")]` in the mutex implementations.

File: core/src/parking_lot.rs (hunk 2 above)
Comment (arthurprs) [reply core/src/parking_lot.rs]: It's a debug artifact, I'll remove it.

File: core/src/parking_lot.rs (hunk 1 above)
Comment (arthurprs) [reply core/src/parking_lot.rs]: I forgot to comment about this. I was thinking that the actual implementation could do a first-run like this and then do a second pass to confirm (locking all queues).

File: core/src/lib.rs (hunk 4 above)
Comment (arthurprs) [reply core/src/lib.rs]: I have no idea how I came up with these, resources is a textbook parlance. I'll think of something.

Comment (arthurprs) [reply core/src/lib.rs]: Good idea.

Comment (Amanieu) [reply core/src/lib.rs]: I think that with the added submodule and some naming changes, it becomes clearer:

- `parking_lot_core::deadlock_detector::acquire_resource`
- `parking_lot_core::deadlock_detector::release_resource`

**Comment by arthurprs** [comment]
I moved a lot of code around now, looks much better.

What's the most useful way of reporting a deadlock?

**Comment by Amanieu** [comment]
> What's the most useful way of reporting a deadlock?

You should show a list of all the threads involved in the cycle, and print a backtrace for each thread. The backtrace can be done by waking each thread up one by one and using a special flag to indicate that it should print a backtrace and then go back to sleep.

**Comment by arthurprs** [comment]
Do you think eprintln!() is enough for the first cut?

**Comment by Amanieu** [comment]
I think that's probably fine for an initial implementation. Ideally you should return a `struct` which contains information about the cycle and a list of thread backtraces. It should be the job of the caller to actually decide how the information should be printed (or logged, etc).

File: src/deadlock.rs [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+#[inline]
+pub fn acquire_resource(_key: usize) {
+    // noop
+}
+
+#[cfg(not(feature = "deadlock_detection"))]
+#[inline]
+pub fn release_resource(_key: usize) {
+    // noop
+}
```
Comment (Amanieu) [inline src/deadlock.rs]: I meant that you should make these functions no-ops in `parking_lot_core`, not here.

Comment (arthurprs) [reply src/deadlock.rs]: Ahh, makes sense.

File: core/src/parking_lot.rs [hunk 6]
```diff
@@ -1040,112 +1091,253 @@ unsafe fn unpark_filter_internal(key: usize,
...
+    use std::sync::mpsc;
+    use std::sync::atomic::Ordering;
+    use std::collections::HashSet;
+    use backtrace::Backtrace;
+    use petgraph;
+    use petgraph::graphmap::DiGraphMap;
+
+    pub(super) unsafe fn on_unpark(td: &ThreadData) {
+        if td.deadlock_data.deadlocked.get() {
+            if let Some(sender) = (*td.deadlock_data.backtrace_sender.get()).take() {
```
Comment (Amanieu) [inline core/src/parking_lot.rs]: This should never be `None`, so just use `unwrap()` here.

File: core/src/parking_lot.rs [hunk 7]
```diff
@@ -1040,112 +1091,253 @@ unsafe fn unpark_filter_internal(key: usize,
...
+            }
+        }
+
+        for b in &(*table).entries[..] {
+            b.mutex.unlock();
+        }
+
+        // find cycles
+        let cycles = graph_cycles(&graph);
+        // println!("cycles {:?}", cycles);
```
Comment (Amanieu) [inline core/src/parking_lot.rs]: Delete commented-out code.

File: core/src/parking_lot.rs [hunk 8]
```diff
@@ -1040,112 +1091,253 @@ unsafe fn unpark_filter_internal(key: usize,
...
-            // double check with a full hash table lock
-            check_wait_graph_cycle(true)
-        } else {
-            false
+        unsafe {
+            // fast pass
+            if check_wait_graph_fast() {
+                // double check
+                let deadlocks = check_wait_graph_slow();
+                // TODO: call a configurable callback
```
Comment (Amanieu) [inline core/src/parking_lot.rs]: Rather than using a callback, I would suggest returning a `struct` with all the backtraces of deadlocked threads.

File: core/src/parking_lot.rs [hunk 9]
```diff
@@ -1040,112 +1091,253 @@ unsafe fn unpark_filter_internal(key: usize,
...
+        deadlock_impl::release_resource(_key);
+    }
+
+    /// TODO: doc
+    #[cfg(feature = "deadlock_detection")]
+    #[inline]
+    pub fn check_deadlock() -> bool {
+        deadlock_impl::check_deadlock()
+    }
+}
```
Comment (Amanieu) [inline core/src/parking_lot.rs]: `acquire_resource` and `release_resource` should be unsafe functions. Consider what happens if we get a false-positive deadlock: the thread will be woken up and then will get stuck in `on_deadlock`. If it is unparked again, it will panic, which will leave the bucket in an inconsistent state.

File: core/src/lib.rs [hunk 10]
```diff
@@ -46,6 +46,10 @@ extern crate rand;
 
 #[cfg(feature = "deadlock_detection")]
 extern crate petgraph;
+#[cfg(feature = "deadlock_detection")]
+extern crate thread_id;
```
[the remark this answers is not in the document]
Comment (arthurprs) [reply core/src/lib.rs]: The thing is that the std::thread IDs are not the system TIDs. I think in this case reporting the system TIDs is important to allow using further debug using gdb etc.

File: core/src/parking_lot.rs (hunk 9 above)
Comment (arthurprs) [reply core/src/parking_lot.rs]: Good point.

File: core/src/lib.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (Amanieu) [reply core/src/lib.rs]: Ok, fair enough.

File: core/src/parking_lot.rs (hunk 8 above)
Comment (arthurprs) [reply core/src/parking_lot.rs]: Indeed, it's also easier.

File: src/deadlock.rs [hunk 11]
```diff
@@ -1,4 +1,38 @@
...
+//!
+//! # Example
+//!
+//! ```
+//! #![feature(deadlock_detection)]
+//! use std::thread;
+//! use std::time::Duration;
+//! use parking_lot::deadlock;
+//!
+//! thread::spawn(move || {
```
Comment (Amanieu) [inline src/deadlock.rs]: Add a comment here: `// Create a background thread which checks for deadlocks every 10s`

**Comment by Amanieu** [comment]
Ok, I think that's all. Just update that comment and I will merge it.

Thanks for all the great work! :+1:

**Comment by arthurprs** [comment]
Done, should we do something about MutexGuard being Send'able?

**Comment by Amanieu** [comment]
I really don't want to break the API for this, but it seems that deadlock detection is fundamentally impossible if we allow lock guards to be sent to other threads.

One alternative would be to make the deadlock detection resilient to false positives and simply add a disclaimer that deadlock detection may be unreliable if `MutexGuard`s are sent between threads. However this is tricky to implement since the current `ThreadParker` implementation is only designed to handle a single unparking thread.

**Comment by arthurprs** [comment]
Let's add a disclaimer for now and in the future we can make the MutexGuard *not* Send'able when deadlock_detection is enabled.

**Comment by arthurprs** [comment]
It wasn't as easy but I got it working.

File: src/deadlock.rs [hunk 12]
```diff
@@ -33,10 +37,17 @@
...
+use std::marker::PhantomData;
+
 #[cfg(feature = "deadlock_detection")]
 pub use parking_lot_core::deadlock::check_deadlock;
 pub(crate) use parking_lot_core::deadlock::{acquire_resource, release_resource};
 
+#[cfg(not(feature = "deadlock_detection"))]
+pub(crate) struct DeadlockDetectionMarker(PhantomData<()>);
+#[cfg(feature = "deadlock_detection")]
+pub(crate) struct DeadlockDetectionMarker(PhantomData<*const ()>); // !Send
```
Comment (arthurprs) [inline src/deadlock.rs]: I made it a struct other than a type alias so that the user can get `DeadlockDetectionMarker` in the !Send error message.

Comment (Amanieu) [reply src/deadlock.rs]: Can you check that `MutexGuard<T>` is still `Sync` when `T: Sync`? Add a test for it similar to the `Send` test for the guards.

File: .travis.yml [hunk 13]
```diff
@@ -32,7 +34,8 @@ after_success:
 
 env:
   global:
-  - TRAVIS_CARGO_NIGHTLY_FEATURE=nightly
+  - TRAVIS_CARGO_NIGHTLY_FEATURE=nightly,deadlock_detection
```
Comment (Amanieu) [inline .travis.yml]: TRAVIS_CARGO_NIGHTLY_FEATURE should be left as it was before. The `deadlock_detection` feature should be explicitly tested separately.

File: .travis.yml [hunk 14]
```diff
@@ -19,6 +19,8 @@ script:
 - travis-cargo build
 - travis-cargo test
 - travis-cargo doc -- --no-deps -p parking_lot -p parking_lot_core
+- travis-cargo --only nightly test -- --all-features
+- travis-cargo --only nightly doc -- --all-features --no-deps -p parking_lot -p parking_lot_core
```
Comment (Amanieu) [inline .travis.yml]: Because of the way TRAVIS_CARGO_NIGHTLY_FEATURE is set, both tests are done with `deadlock_detection` entabled.

File: benchmark/Cargo.toml [hunk 15]
```diff
@@ -18,3 +18,4 @@ path = "src/rwlock.rs"
 
 [features]
 nightly = ["parking_lot/nightly"]
+deadlock_detection = ["parking_lot/deadlock_detection"]
```
Comment (Amanieu) [inline benchmark/Cargo.toml]: The benchmarks don't need deadlock detection.

File: .travis.yml [hunk 16]
```diff
@@ -18,7 +18,9 @@ script:
 - cd ..;
 - travis-cargo build
 - travis-cargo test
+- travis-cargo test -- --features=deadlock_detection
 - travis-cargo doc -- --no-deps -p parking_lot -p parking_lot_core
+- travis-cargo --only nightly doc -- --all-features --no-deps -p parking_lot -p parking_lot_core
```
Comment (Amanieu) [inline .travis.yml]: Why was this line added?

File: benchmark/Cargo.toml (hunk 15 above)
Comment (arthurprs) [reply benchmark/Cargo.toml]: It's not enabled by default, it's useful to measure the impact.

File: .travis.yml (hunk 16 above)
Comment (arthurprs) [reply .travis.yml]: Ahh, it should have replaced the one above

File: benchmark/Cargo.toml (hunk 15 above)
Comment (Amanieu) [reply benchmark/Cargo.toml]: Fair enough

**Comment by Amanieu** [comment]
Do a `cargo fmt` to clean up the code style (for both parking_lot and parking_lot_core), and it should be good to go.

File: .travis.yml (hunk 16 above)
Comment (arthurprs) [reply .travis.yml]: It's to build the docs including all features, otherwise deadlock_detection will not show.

Comment (Amanieu) [reply .travis.yml]: Then just add `--all-features` to the cargo doc line above.

**Comment by arthurprs** [comment]
I squashed the deadlock detection changes, rustfmt changed quite a lot.

**Comment by Amanieu** [comment]
Thanks for all the hard work! :+1:
